### PR TITLE
feat(agent): two-way communications

### PR DESCRIPTION
**Pull request by andrewazores**
* chore(svc): extract EventOptionsBuilder to -core and use new CryostatFlightRecorderService

* add unimplemented overrides

* test(smoketest): enable API writes on one agent-equipped sample app

* chore(serial): extract recording descriptor to -core

* chore(activerecordings): clean up an error handler

* feat(agent): implement dynamic start of JFR over HTTP

* bump -core version

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
* [ ] Read the [contributing guidelines](https://github.com/cryostatio/cryostat/blob/main/CONTRIBUTING.md)
* [ ] Linked a relevant issue which this PR resolves
* [ ] Linked any other relevant issues, PR's, or documentation, if any
* [ ] Resolved all conflicts, if any
* [ ] Rebased your branch PR on top of the latest upstream `main` branch
* [ ] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [ ] [Signed all commits using a GPG signature](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#gpg-commit-signature-verification)

**To recreate commits with GPG signature** `git fetch upstream && git rebase --force --gpg-sign upstream/main`
_______________________________________________

Fixes: #<issue number>

## Description of the change:
*This change adds allows a match expression example to be copied to the clipboard...*

## Motivation for the change:
*This change is helpful because users may want to copy the example for easier use...*

## How to manually test:
1. *Run CRYOSTAT_IMAGE=quay.io... sh smoketest.sh...*
2. *...*

### CODE REVIEW

**Comment by andrewazores** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6175405282

**Comment by andrewazores** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6176094283

**Review by mwangggg — DISMISSED** [review]
Looks good to me :+1:

**Comment by andrewazores** [comment]
Stopping running recordings looks possibly bugged:

```
SEVERE: HTTP 500: No value present
io.vertx.ext.web.handler.HttpException: Internal Server Error
Caused by: java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:143)
	at io.cryostat.recordings.RecordingTargetHelper.lambda$stopRecording$6(RecordingTargetHelper.java:296)
	at io.cryostat.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:146)
	at io.cryostat.recordings.RecordingTargetHelper.stopRecording(RecordingTargetHelper.java:275)
	at io.cryostat.recordings.RecordingTargetHelper.stopRecording(RecordingTargetHelper.java:268)
	at io.cryostat.net.web.http.api.v1.TargetRecordingPatchStop.handle(TargetRecordingPatchStop.java:40)
	at io.cryostat.net.web.http.api.v1.TargetRecordingPatchHandler.handleAuthenticated(TargetRecordingPatchHandler.java:105)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:81)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:51)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

@aali309 also saw this during a similar workflow:

```
Sep 14, 2023 7:31:37 PM org.slf4j.impl.JDK14LoggerAdapter fillCallerData
WARNING: Thread Thread[vert.x-eventloop-thread-2,5,main] has been blocked for 237060 ms, time limit is 2000 ms
io.vertx.core.VertxException: Thread blocked
	at java.base@17.0.8/jdk.internal.misc.Unsafe.park(Native Method)
	at java.base@17.0.8/java.util.concurrent.locks.LockSupport.park(LockSupport.java:211)
	at java.base@17.0.8/java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1864)
	at java.base@17.0.8/java.util.concurrent.ForkJoinPool.unmanagedBlock(ForkJoinPool.java:3465)
	at java.base@17.0.8/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3436)
	at java.base@17.0.8/java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1898)
	at java.base@17.0.8/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2072)
	at app//io.cryostat.net.AgentJFRService.getAvailableRecordings(AgentJFRService.java:124)
	at app//io.cryostat.recordings.RecordingArchiveHelper.getDescriptorByName(RecordingArchiveHelper.java:997)
	at app//io.cryostat.recordings.RecordingArchiveHelper.lambda$saveRecording$1(RecordingArchiveHelper.java:375)
	at app//io.cryostat.recordings.RecordingArchiveHelper$$Lambda$1639/0x00007fcf1cb846a0.execute(Unknown Source)
	at app//io.cryostat.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:146)
	at app//io.cryostat.recordings.RecordingArchiveHelper.saveRecording(RecordingArchiveHelper.java:371)
	at app//io.cryostat.recordings.RecordingTargetHelper.lambda$scheduleRecordingTasks$16(RecordingTargetHelper.java:624)
	at app//io.cryostat.recordings.RecordingTargetHelper$$Lambda$1636/0x00007fcf1cb84000.handle(Unknown Source)
	at app//io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at app//io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
	at app//io.vertx.core.impl.future.FutureBase$$Lambda$433/0x00007fcf1c633838.run(Unknown Source)
	at app//io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at app//io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at app//io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at app//io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:403)
	at app//io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at app//io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@17.0.8/java.lang.Thread.run(Thread.java:833)
```

**Comment by andrewazores** [comment]
After navigating away (to Events) and back to Recordings, I have this:

image

Definitely bugged.

**Comment by andrewazores** [comment]
I have a hunch that the request sent to PATCH the recording on the Agent side ends up both stopping the recording as well as erasing its name, which then breaks the server's ability to look that recording up again since it expects it to still be there with the same name.

**Comment by aali309** [comment]
1. recording on a different Agent/HTTP target fails as expected
 2. recording on quarkus-test-agent-2 (http://localhost:9988) works as expected with one recording
 3. creating another recording produces some errors when stopping active active recordings. I think it is an issue having two recordings ( tried like 4 times just to be sure) and I get the same results. 
```Sep 14, 2023 7:42:41 PM org.slf4j.impl.JDK14LoggerAdapter fillCallerData
WARNING: Thread Thread[vert.x-eventloop-thread-2,5,main] has been blocked for 900134 ms, time limit is 2000 ms
io.vertx.core.VertxException: Thread blocked
	at java.base@17.0.8/jdk.internal.misc.Unsafe.park(Native Method)
	at java.base@17.0.8/java.util.concurrent.locks.LockSupport.park(LockSupport.java:211)
	at java.base@17.0.8/java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1864)
	at java.base@17.0.8/java.util.concurrent.ForkJoinPool.unmanagedBlock(ForkJoinPool.java:3465)
	at java.base@17.0.8/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3436)
	at java.base@17.0.8/java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1898)
	at java.base@17.0.8/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2072)
	at app//io.cryostat.net.AgentJFRService.getAvailableRecordings(AgentJFRService.java:124)
	at app//io.cryostat.recordings.RecordingArchiveHelper.getDescriptorByName(RecordingArchiveHelper.java:997)
	at app//io.cryostat.recordings.RecordingArchiveHelper.lambda$saveRecording$1(RecordingArchiveHelper.java:375)
	at app//io.cryostat.recordings.RecordingArchiveHelper$$Lambda$1639/0x00007fcf1cb846a0.execute(Unknown Source)
	at app//io.cryostat.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:146)
	at app//io.cryostat.recordings.RecordingArchiveHelper.saveRecording(RecordingArchiveHelper.java:371)
	at app//io.cryostat.recordings.RecordingTargetHelper.lambda$scheduleRecordingTasks$16(RecordingTargetHelper.java:624)
	at app//io.cryostat.recordings.RecordingTargetHelper$$Lambda$1636/0x00007fcf1cb84000.handle(Unknown Source)
	at app//io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at app//io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
	at app//io.vertx.core.impl.future.FutureBase$$Lambda$433/0x00007fcf1c633838.run(Unknown Source)
	at app//io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at app//io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at app//io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at app//io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:403)
	at app//io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at app//io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@17.0.8/java.lang.Thread.run(Thread.java:833)
```
Also crashes on the web part on `https://localhost:8181/recordings` with no success to reload/refresh 
image

Working with one recording nothing breaks but:
1. Deleting the recording works as expected
2. Stopping the recording produces 

`Danger alert:Request failed (500 Internal Server Error)
No value present caused by NoSuchElementException: No value present`

3. and the Deleting stops working as expected. @andrewazores suspects its because the name of the recording got changed during the stop operation which is i verify to be true because when the page is reloaded the recording was renamed to `unnamed Recording` and then Deleting works. 
image

4. Downloading and view Report works as expected (with Grafana as well).
5. Creating snapshot recordings also works as expected

**Comment by aali309** [comment]
Everything looking great including testing with multiple recordings + snapshots.

**Review by aali309 — APPROVED** [review]

**Comment by andrewazores** [comment]
```
INFO: Outgoing WS message: {"meta":{"category":"ActiveRecordingStopped","type":{"type":"application","subType":"json"},"serverTime":1695129660},"message":{"target":"service:jmx:rmi:///jndi/rmi://cryostat:9097/jmxrmi","recording":{"downloadUrl":"https://localhost:8181/api/v1/targets/service:jmx:rmi:%2F%2F%2Fjndi%2Frmi:%2F%2Fcryostat:9097%2Fjmxrmi/recordings/cryostat-agent","reportUrl":"https://localhost:8181/api/v1/targets/service:jmx:rmi:%2F%2F%2Fjndi%2Frmi:%2F%2Fcryostat:9097%2Fjmxrmi/reports/cryostat-agent","metadata":{"labels":{}},"archiveOnStop":false,"id":1,"name":"cryostat-agent","state":"STOPPED","startTime":1695129573590,"duration":0,"continuous":true,"toDisk":true,"maxSize":0,"maxAge":60000}}}
Sep 19, 2023 1:21:00 PM io.cryostat.core.log.Logger error
SEVERE: HTTP 500: No value present
io.vertx.ext.web.handler.HttpException: Internal Server Error
Caused by: java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:143)
	at io.cryostat.recordings.RecordingTargetHelper.lambda$stopRecording$6(RecordingTargetHelper.java:300)
	at io.cryostat.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:151)
	at io.cryostat.recordings.RecordingTargetHelper.stopRecording(RecordingTargetHelper.java:279)
	at io.cryostat.recordings.RecordingTargetHelper.stopRecording(RecordingTargetHelper.java:272)
	at io.cryostat.net.web.http.api.v1.TargetRecordingPatchStop.handle(TargetRecordingPatchStop.java:40)
	at io.cryostat.net.web.http.api.v1.TargetRecordingPatchHandler.handleAuthenticated(TargetRecordingPatchHandler.java:105)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:80)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:50)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

The stoppage notification is emitted, but after navigating away and back to the Recordings page and triggering a re-query of data, the recording state is actually still running. That matches what would be expected from the stack trace. This behaviour is also present on `main`.

**Comment by andrewazores** [comment]
Agent-side logs:

```
2023-09-19 13:29:59,648 INFO  [io.cry.age.Harvester] (RMI TCP Connection(12)-10.0.2.100) cryostat-agent(3) STOPPED
2023-09-19 13:29:59,649 INFO  [io.cry.age.Harvester] (RMI TCP Connection(12)-10.0.2.100) cryostat-agent(3) CLOSED
2023-09-19 13:29:59,838 INFO  [io.cry.age.CryostatClient] (cryostat-agent-worker-0) POST 200 (quarkus-test-agent-1_default_20230919T132959Z.jfr -> https://localhost:8181/api/beta/recordings/KnY4JVq_p8hVVoh-GsmOAarLxC6bwdSX9P2qDlU0520=): 392 KB/PT0.188978532S
2023-09-19 13:29:59,851 INFO  [io.cry.age.Harvester] (cryostat-agent-harvester) cryostat-agent(4) RUNNING
2023-09-19 13:30:03,031 INFO  [io.cry.age.CryostatClient] (cryostat-agent-registration) GET https://localhost:8181/api/v2.2/discovery/d5535a73-4774-40f4-be79-4721ef5787f2?token=eyJjdHkiOiJKV1QiLCJlbmMiOiJBMjU2R0NNIiwiYWxnIjoiZGlyIn0..ZupKK7gWmaotvvF_.EuQ2znPibxbWpN2a9NQjTIR3De8PPlglQ0aDz9A97MJnFDKWAM1-UQ2ohLPZxiiE5we7DOloGk9rTIsSdGgQkn_RHSkQQzLDCjk1ZJaHQa566LoNA4ZlpmRZspYq6WFvaY3BFobeytzgmAiFTupahelVap-yM0g3CWvKQNTNgKWOS5g-bUMZVNZHXBFnlOLgBb8ZV_HKO57IyhdwIR_nl8QSUJ2LykBtGOiQvZVat0sHO00NiIB__D-w1v1IXo4BpLHPPrAj14o3E_9B-ZV6AOv35HNmUg_ftYnZmhtFO0oBh3UxAov1gAx2K_Xt5mIzdL2Pf6NmXsNNno6BC_YF17oRdOb-vZke3_zFhqmZWIiBWRQqGGRXFyYBVXW45YL7tdVAMh-kprt5V91_9Ok0OBf66pdsTye9NBn8lu72klKc6seAoVOyAmL_wRlftOr_9e3ei_QYA4XL8aXsMU2UWdwn2KWojwkx6dWGJES4BpV55EqHonok4b7oiqmYonD43weHv3qKzWoCzyn0dBSJS40p4FkP7NZWDJS1d-EOUZvH5wYa1pIJ-U6FYkQ4AB37n6Uqrb4d154u.oPDrIfTyxrzX_jenTxZkrQ HTTP/1.1
2023-09-19 13:30:03,054 INFO  [io.cry.age.CryostatClient] (cryostat-agent-worker-2) GET https://localhost:8181/api/v2.2/discovery/d5535a73-4774-40f4-be79-4721ef5787f2?token=eyJjdHkiOiJKV1QiLCJlbmMiOiJBMjU2R0NNIiwiYWxnIjoiZGlyIn0..ZupKK7gWmaotvvF_.EuQ2znPibxbWpN2a9NQjTIR3De8PPlglQ0aDz9A97MJnFDKWAM1-UQ2ohLPZxiiE5we7DOloGk9rTIsSdGgQkn_RHSkQQzLDCjk1ZJaHQa566LoNA4ZlpmRZspYq6WFvaY3BFobeytzgmAiFTupahelVap-yM0g3CWvKQNTNgKWOS5g-bUMZVNZHXBFnlOLgBb8ZV_HKO57IyhdwIR_nl8QSUJ2LykBtGOiQvZVat0sHO00NiIB__D-w1v1IXo4BpLHPPrAj14o3E_9B-ZV6AOv35HNmUg_ftYnZmhtFO0oBh3UxAov1gAx2K_Xt5mIzdL2Pf6NmXsNNno6BC_YF17oRdOb-vZke3_zFhqmZWIiBWRQqGGRXFyYBVXW45YL7tdVAMh-kprt5V91_9Ok0OBf66pdsTye9NBn8lu72klKc6seAoVOyAmL_wRlftOr_9e3ei_QYA4XL8aXsMU2UWdwn2KWojwkx6dWGJES4BpV55EqHonok4b7oiqmYonD43weHv3qKzWoCzyn0dBSJS40p4FkP7NZWDJS1d-EOUZvH5wYa1pIJ-U6FYkQ4AB37n6Uqrb4d154u.oPDrIfTyxrzX_jenTxZkrQ : 200
```

It looks like what is happening is that the remote JMX connection invokes the `stopRecording` operation, which the target JVM accepts and performs. Immediately, the Agent Harvester code executing within the target JVM sees this updated condition and reacts:

https://github.com/cryostatio/cryostat-agent/blob/8db9b4cf177781090baec976d924fc87533f82ec/src/main/java/io/cryostat/agent/Harvester.java#L193

The Agent dumps the stopped recording contents to a file, closes the recording, uploads the file dump to the server, and then starts a new replacement recording with the same name and other settings according to the Harvester configuration. The server meanwhile, after sending the JMX stopRecording command:

```java
                    Optional<IRecordingDescriptor> descriptor =
                            getDescriptorByName(connection, recordingName);
                    if (descriptor.isPresent()) {
                        IRecordingDescriptor d = descriptor.get();
                        if (d.getState().equals(RecordingState.STOPPED) && quiet) {
                            return d;
                        }
                        connection.getService().stop(d);
                        this.cancelScheduledTasksIfExists(targetId, recordingName);
                        HyperlinkedSerializableRecordingDescriptor linkedDesc =
                                new HyperlinkedSerializableRecordingDescriptor(
                                        d,
                                        webServer.get().getDownloadURL(connection, d.getName()),
                                        webServer.get().getReportURL(connection, d.getName()),
                                        RecordingState.STOPPED);
                        this.issueNotification(targetId, linkedDesc, STOP_NOTIFICATION_CATEGORY);
                        return getDescriptorByName(connection, recordingName).get();
                    } else {
                        throw new RecordingNotFoundException(targetId, recordingName);
                    }
```

does a little other work of its own, then tries to query the target for the updated state of the recording by the same name so that it can inform the original client of the latest state. The expectation should be that only the `running -> stopped` state has changed, but this re-query is done to ensure that we really report the latest state to the client. I believe the `NoSuchElementException` arises because this follow-up query happens to occur in the timing interval where the Agent has deleted the original recording and has not yet re-started its replacement, so there is actually no recording with the expected name in the target JVM anymore.

tl;dr this is a conflict between trying to use manual recording stop operations on the Agent Harvester-managed recording. I'm not sure if it's really a bug in that case but maybe there is something to be done to make this result a little more clear and explicit to the user. In the Agent HTTP connection case that is probably easier to do since we have more control. In this JMX case, I don't know what can be done really.
